x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -332,6 +353,135 @@ private void createApiKeyAndIndexIt(
         }));
     }
 
+    public void updateApiKey(
+        Authentication authentication,
+        UpdateApiKeyRequest request,
+        Set<RoleDescriptor> userRoles,
+        ActionListener<UpdateApiKeyResponse> listener
+    ) {
+        ensureEnabled();
+
+        if (authentication == null) {
+            listener.onFailure(new IllegalArgumentException("authentication must be provided"));
+            return;
+        }
+
+        findVersionedApiKeyDocsForSubject(authentication, new String[] { request.getId() }, ActionListener.wrap((versionedDocs) -> {
+            final var apiKeyId = request.getId();
+
+            if (versionedDocs.isEmpty()) {
+                throw apiKeyNotFound(apiKeyId);
+            }
+
+            validateCurrentApiKeyDocForUpdate(apiKeyId, authentication, single(versionedDocs).doc());
+
+            doBulkUpdate(
+                buildBulkUpdateRequest(versionedDocs, authentication, request, userRoles),
+                ActionListener.wrap(bulkResponse -> translateResponseAndClearCache(apiKeyId, bulkResponse, listener), listener::onFailure)
+            );
+        }, listener::onFailure));
+    }
+
+    // package-private for testing
+    void validateCurrentApiKeyDocForUpdate(String apiKeyId, Authentication authentication, ApiKeyDoc apiKeyDoc) {
+        assert authentication.getEffectiveSubject().getUser().principal().equals(apiKeyDoc.creator.getOrDefault("principal", null));
+        assert authentication.getEffectiveSubject().getRealm().getName().equals(apiKeyDoc.creator.getOrDefault("realm", null));
+
+        if (Version.fromId(apiKeyDoc.version).before(Version.V_8_2_0)) {
+            throw new ValidationException().addValidationError(
+                "cannot update legacy api key [" + apiKeyId + "] with version [" + Version.fromId(apiKeyDoc.version) + "]"
+            );
+        }
+        if (isActive(apiKeyDoc) == false) {
+            throw new ValidationException().addValidationError("cannot update inactive api key [" + apiKeyId + "]");
+        }
+        if (Strings.isNullOrEmpty(apiKeyDoc.name)) {
```

Comment:
I don't think we should support auto-upgrading API keys without names -- we have an implicit assumption that an 8.x API key has a name. Auto-generating a name or using the ID as the name seems like too much of a BWC overreach. Let me know if you feel strongly this.